### PR TITLE
module_3

### PR DESCRIPTION
**Pull request by AGrandeC**
Adjunto mi solución del módulo 3. Con los módulos 2 y 3, pese a que los trabajé antes de clase por mi cuenta, la solución difería tanto de la presentada en clase que me parecía una mierda (hablando mal y pronto jajaja)

Ambos módulos los trabajé también después de ver la solución, de hecho, mi solución al módulo 2 la terminé borrando y utilizaba casi todo lo que se presentó en clase (por eso no la voy a subir ya que no tiene sentido)

Guille en clase comentó que era la esperable así que me lanzo a subir esta PR con lo que hice en su día.

Al menos he aprendido y estoy aplicando los nuevos conceptos para los módulos 4 y 5 (WIP)

PS: ya comentamos en clase que el undersampling no tenía sentido, que debería haber usado predict_proba, hacer un baseline más sencillo aún (quitándo las categóricas por ejemplo), añadir un pipeline para construir el modelo, el train-test split pese haber pensado que importaba la fecha (de hecho lo escribí) al final lo hice sólo por order_id...

### CODE REVIEW

File: src/module_3/module3_statistical_learning.md [hunk 1]
```diff
@@ -0,0 +1,1368 @@
...
+  </tbody>
+</table>
+<p>2880549 rows × 27 columns</p>
+</div>
+
+
+
+
+```python
+items_ordered = df_base[df_base['outcome'] == 1].groupby('order_id')['variant_id'].count().reset_index().rename(columns={"variant_id": "order_size"})
```
Comment (gsantam) [inline src/module_3/module3_statistical_learning.md:437]: the name is not great, what about: `order_size_df`

Comment (gsantam) [reply src/module_3/module3_statistical_learning.md:437]: also you can do: `df_base[df_base['outcome'] == 1].groupby('order_id').size().reset_index()` instead

File: src/module_3/module3_statistical_learning.md [hunk 2]
```diff
@@ -0,0 +1,1368 @@
+```python
+import os
+import numpy as np
+import pandas as pd
+from summarytools import dfSummary
```
Comment (gsantam) [inline src/module_3/module3_statistical_learning.md:5]: is this a python module? Never heard of it, but doesn't seem to use the best naming convention 😆

File: src/module_3/module3_statistical_learning.md [hunk 3]
```diff
@@ -0,0 +1,1368 @@
...
+    </tr>
+  </tbody>
+</table>
+</div>
+
+
+
+
+```python
+dfSummary(df_orders)
```
Comment (gsantam) [inline src/module_3/module3_statistical_learning.md:661]: the output of this function is very confusing, there are much better things to do summary of data frames, you can start by a simple `describe` or can go to more sophisticated things like `from pandas_profiling import ProfileReport`

File: src/module_3/module3_statistical_learning.md [hunk 4]
```diff
@@ -0,0 +1,1368 @@
...
+      <td id="T_7f732_row27_col4" class="data row27 col4" ><img src = "data:image/png;base64, iVBORw0KGgoAAAANSUhEUgAAAKoAAABGCAYAAABc8A97AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8g+/7EAAAACXBIWXMAAA9hAAAPYQGoP6dpAAABYUlEQVR4nO3doU7DUBiA0UEQU6Rifh7TNyA8w8IzTu4ZeJV5xORQQxEIZqSFsG85R7c3FV9yW/H33pxOpwVcutv/fgD4CaGScHfugvV6/TQMw2rK4ofD4XW/379MuRe+OhvqMAyrzWbzNmXx3W43KXD4ztZPglBJECoJQiVBqCQIlQShkiBUEoRKglBJECoJQiVBqCQIlQShkiBUEoRKglBJECoJQiVBqCQIlQShkiBUEoRKglBJECoJQiVBqCQIlQShkiBUEoRKglBJECoJQiXh7GETcxyPx4dxHJ+n3u9UFT78aajL5fJ+6okqi4VTVfhk6ydBqCQIlQShkiBUEoRKglBJECoJQiVBqCQIlQShkiBUEoRKglBJECoJQiVBqCQIlQShkvCnw31zzZliNcF6XS461DlTrCZYr4utnwShkiBUEi76HXUOvxO6Llcb6tzfCW2328dxHCd9kIn8970DYCVFi+c6a9sAAAAASUVORK5CYII="></img></td>
+      <td id="T_7f732_row27_col5" class="data row27 col5" >0<br>(0.0%)</td>
+    </tr>
+  </tbody>
+</table>
+
+
+
+
+# Milestone 1
```
Comment (gsantam) [inline src/module_3/module3_statistical_learning.md:959]: could you briefly explain what Milestone 1 is and what are you gonna do?

File: src/module_3/module3_statistical_learning.md [hunk 5]
```diff
@@ -0,0 +1,1368 @@
...
+
+    
+![png](module3_statistical_learning_files/module3_statistical_learning_9_0.png)
+    
+
+
+## Create train and test sets
+
+Puntos a tener en cuenta
+- Hacer el split por order_id - No puede haber un order_id en train y test
```
Comment (gsantam) [inline src/module_3/module3_statistical_learning.md:1058]: quieres decir "un mismo `order_id`?

Comment (gsantam) [reply src/module_3/module3_statistical_learning.md:1058]: put column names with format: `order_id`

File: src/module_3/module3_statistical_learning.md [hunk 6]
```diff
@@ -0,0 +1,1368 @@
...
+# Before splitting the data we'll handle datetimes
+
+### Extract components of datetimes
+df_orders = df_orders.drop('created_at', axis=1) # I'll only keep order_date as they display almost the same info
+
+df_orders['order_year'] = df_orders['order_date'].dt.year
+df_orders['order_month'] = df_orders['order_date'].dt.month
+df_orders['order_day_of_week'] = df_orders['order_date'].dt.dayofweek 
+
+df_orders = df_orders.drop('order_date', axis=1)
```
Comment (gsantam) [inline src/module_3/module3_statistical_learning.md:1072]: what's the point of dropping columns?

File: src/module_3/module3_statistical_learning.md [hunk 7]
```diff
@@ -0,0 +1,1368 @@
...
+
+    # Split order IDs into train and test groups 
+    train_order_ids, test_order_ids = train_test_split(unique_order_ids, 
+                                                       test_size=test_size, 
+                                                       random_state=random_state)
+
+    # Optionally, split train order IDs further into train and validation groups
+    if validation_size > 0:
+        train_order_ids, validation_order_ids = train_test_split(train_order_ids,
+                                                                 test_size=validation_size / (1 - test_size),
```
Comment (gsantam) [inline src/module_3/module3_statistical_learning.md:1105]: this doesn't work, if `validation_size = 0.1` and `test_size = 0.1`, `validation_size / (1 - test_size) = 0.1/0.9 = 0.1111` so you are having a different size than 0.1 which is the one you passed in the parameter for the validation

File: src/module_3/module3_statistical_learning.md [hunk 8]
```diff
@@ -0,0 +1,1368 @@
...
+                                                                 random_state=random_state)
```
Comment (gsantam) [inline src/module_3/module3_statistical_learning.md:1106]: this functions are a bit complicated. I recommend you that after you check that `len(test_order_ids) == int(len(unique_order_ids)* test_size ` and  `len(validation_order_ids) == int(len(validation_order_ids)* validation_size `

File: src/module_3/module3_statistical_learning.md [hunk 9]
```diff
@@ -0,0 +1,1368 @@
...
+
+    # Create split masks
+    train_mask = df['order_id'].isin(train_order_ids)
+    validation_mask = df['order_id'].isin(validation_order_ids)
+    test_mask = df['order_id'].isin(test_order_ids)
+
+    # Split the DataFrame based on masks
+    train_df = df[train_mask]
```
Comment (gsantam) [inline src/module_3/module3_statistical_learning.md:1114]: this validation is not correct thought, you are putting orders in test which may happen before train and viceversa so you are not respecting the temporal evaluation and therefore you have very big data leakages

File: src/module_3/module3_statistical_learning.md [hunk 10]
```diff
@@ -0,0 +1,1368 @@
...
+    pickle.dump(encoders, f)
+
+# ------------------ Model Training ------------------
+    
+X_train = train_df.drop(label_col, axis=1)
+y_train = train_df[label_col]
+
+# Fit the LogisticRegression model
+log_reg = LogisticRegression() 
+log_reg.fit(X_train, y_train)
```
Comment (gsantam) [inline src/module_3/module3_statistical_learning.md:1148]: you are not scaling the numeric features! this is quite bad if you are using logistic regression

Comment (gsantam) [reply src/module_3/module3_statistical_learning.md:1148]: also you are not removing very correlated features

File: src/module_3/module3_statistical_learning.md [hunk 11]
```diff
@@ -0,0 +1,1368 @@
...
+log_reg = LogisticRegression() 
+log_reg.fit(X_train, y_train) 
+
+# ------------------ Classification ------------------
+
+# Encode test data
+with open(data_path + 'label_encoders.pkl', 'rb') as f:
+    encoders = pickle.load(f)
+
+for col in categorical_cols:
```
Comment (gsantam) [inline src/module_3/module3_statistical_learning.md:1156]: put this into a function or even better a pipeline: https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html

File: src/module_3/module3_statistical_learning.md [hunk 12]
```diff
@@ -0,0 +1,1368 @@
...
+    train_df = df[train_mask]
+    validation_df = df[validation_mask]
+    test_df = df[test_mask]
+
+    return train_df, validation_df, test_df
+
+train_df_0, validation_df_0, test_df_0 = split_by_order_id(df_orders)
+```
+
+### Create a baseline model
```
Comment (gsantam) [inline src/module_3/module3_statistical_learning.md:1123]: technically in a baseline model you don't use ML, it needs to be more simple

File: src/module_3/module3_statistical_learning.md [hunk 13]
```diff
@@ -0,0 +1,1368 @@
...
+```python
+def model_performance_metrics(y_test, y_pred):
+    accuracy = accuracy_score(y_test, y_pred)
+    conf_matrix = confusion_matrix(y_test, y_pred, labels=[1, 0])
+    precision = precision_score(y_test, y_pred, labels=[1, 0])
+    recall = recall_score(y_test, y_pred, labels=[1, 0])
+    f1 = f1_score(y_test, y_pred)
+    roc_auc = roc_auc_score(y_test, y_pred)
+
+    print(f'Accuracy: {accuracy}')
```
Comment (gsantam) [inline src/module_3/module3_statistical_learning.md:1177]: accuracy / precision and recall do not make sense if they don't come with a threshold and therefore I really discourage them as "raw" metrics of a classifier. Instead use ROC AUC or PR AUC

File: src/module_3/module3_statistical_learning.md [hunk 14]
```diff
@@ -0,0 +1,1368 @@
...
+```
+
+    Accuracy: 0.9854730694473871
+    Confusion Matrix:
+     [[     0   3154]
+     [     0 213960]]
+    Precision: 0.0
+    Recall: 0.0
+    F1-Score: 0.0
+    ROC AUC: 0.5
```
Comment (gsantam) [inline src/module_3/module3_statistical_learning.md:1194]: this basically means that your model is random. A logistic regression is not a great model but I really really doubt is not learning anything at all, something is probably off

File: src/module_3/module3_statistical_learning.md [hunk 15]
```diff
@@ -0,0 +1,1368 @@
...
+
+
+Como se podía esperar, el modelo predice sólo 0s ya que las clases están totalmente desbalanceadas. El siguiente paso será aplicar un undersampling para igualar las clases
+
+## Undersampling
+
+En este paso voy a aplicar un undersampling para tener las clases balanceadas al 50%
```
Comment (gsantam) [inline src/module_3/module3_statistical_learning.md:1201]: what's the point of this? As commented in the class undersampling si useless in most of the cases unless your classes are very unbalanced (which is not the case)

File: src/module_3/module3_statistical_learning.md [hunk 16]
```diff
@@ -0,0 +1,1368 @@
...
+
+
+```python
+train_df = train_df_0.copy()
+validation_df = validation_df_0.copy()
+
+### Label Encoder
+encoders = {}
+for col in categorical_cols:
```
Comment (gsantam) [inline src/module_3/module3_statistical_learning.md:1210]: function please!

**Review by gsantam — CHANGES_REQUESTED** [review]
Important issues about the evaluation schema and how do you use linear regression

File: src/module_3/module3_statistical_learning.md (hunk 3 above)
Comment (AGrandeC) [reply src/module_3/module3_statistical_learning.md:661]: will try out `from pandas_profiling import ProfileReport`, first time I've heard of it

File: src/module_3/module3_statistical_learning.md (hunk 6 above)
Comment (AGrandeC) [reply src/module_3/module3_statistical_learning.md:1072]: I think here I tried to replace the date by year, month and day ints because the model didn't let me input dates
